### ISSUE TITLE
ENH: add basic support for M values (parsing, checking for M, accessing values)

### ISSUE THREAD

**Issue by jorisvandenbossche** [shapely/shapely#1648]
With GEOS main (future GEOS 3.12), there is improved support for M values. Those can now be stored under the hood in the CoordinateSequence. The M value isn't used for much (quite similarly as Z values), but we could provide basic functionality to parse/create geometries with M values, check if a geometry has M values and access the M value. 

For example, that would support the workflow of parsing WKB/WKT with M and extracting that value. The parsing now works with GEOS main:

```
>>> geom = shapely.from_wkt("POINT M (1 2 3)")                                                                                                                                                                 
>>> geom                                                                                                                                                                                                       
<POINT M (1 2 3)>
```

But there is currently no way to access this value 3. 

Some possible places to add support for M values:

* Add a ``has_m`` function and class attribute (mimicking `has_z`). We should probably ask upstream to add a `GEOSHasM` to the C API.
* Add an `include_m` keyword in `get_coordinates`. Currently this function defaults to return 2D, but can return 3D if you specify `include_z=True`

**Comment by idanmiara** [comment]
I was wondering how useful is the M coordinate, as I've never seen it actually used in the wild :)

> * Add a `has_m` function and class attribute (mimicking `has_z`). We should probably ask upstream to add a `GEOSHasM` to the C API.

As calling `has_z` and `has_m` in a row can be wasteful, It might be a good idea to also have a function `has_zm` that will either return Nx2 bool array (has z, has m, per geom) or  a N int8 array (first bit=has z, second bit=has m, per geom)

> * Add an `include_m` keyword in `get_coordinates`. Currently this function defaults to return 2D, but can return 3D if you specify `include_z=True`

In the same way, `include_z: int | bool` could be considered (2 bits), although have a confusing name, could be backwards compatible with `include_z: bool` without the addition of extra parameter.

Also we need to consider if there is code where `coord_dim=3` was considered equivalent to `has_z`. which now might not be the case with XYM geometries.

**Comment by mwtoews** [comment]
The M-value is user-defined, but the most common use is to store time. M-values have had long-standing support within several GIS geometry models, so I'd be happy to see it here eventually too. PostGIS has some good inspiration, for example a LineString&nbsp;[Z]M could add timestamps to each coordinate to create a trajectory (see [ST_IsValidTrajectory](https://postgis.net/docs/ST_IsValidTrajectory.html), [ST_DistanceCPA](https://postgis.net/docs/ST_DistanceCPA.html) and [ST_ClosestPointOfApproach](https://postgis.net/docs/ST_ClosestPointOfApproach.html)). Other functions like [ST_SetEffectiveArea](https://postgis.net/docs/ST_SetEffectiveArea.html) return effective areas from each coordinate to the M-values.

**Comment by mwtoews** [comment]
BTW, I've started some of these tasks. Adding `has_z` is trivial. Adding `.m` to `Point` is nearly done. The bigger challenge that I'm looking at (and might need help with) is reworking much of the internal logic related to the 3D coordinates, whether they are XYZ or XYM. For instance https://github.com/shapely/shapely/blob/eeef37fcf398afeb6342d37b63037dc398fa1a5c/src/geos.c#L1007-L1009

The argument `dims` can be 3, but this doesn't distinguish between the two variants. Should this change to have `has_z` and `has_m` parameters instead?

Also the Python signatures, e.g. [`shapely.points`](https://shapely.readthedocs.io/en/2.0.1/reference/shapely.points.html). This could be one of several things:

- non-breaking:
    `points(coords, y=None, z=None, indices=None, out=None, m=None, **kwargs)`
- breaking, possibly yielding unexpected results:
    `points(coords, y=None, z=None, m=None, indices=None, out=None, m=None, **kwargs)`
- breaking, but safer for users to identify fix:
    `points(coords, y=None, *, z=None, m=None, indices=None, out=None, m=None, **kwargs)`

If we are introducing breaking changes for shapely 2.1, should warnings be introduced to the 2.0.* series?

Simpler [`shapely.Point`](https://shapely.readthedocs.io/en/stable/reference/shapely.Point.html) could be refactored something like this:
```python
shapely.Point(1, 2, 3)  # POINT Z (1 2 3)
shapely.Point(1, 2, 3, 4)  # POINT ZM (1 2 3 4)
shapely.Point(1, 2, None, 4)  # POINT M (1 2 4)
```
More complicated [`shapely.LineString`](https://shapely.readthedocs.io/en/stable/reference/shapely.LineString.html):
```python
shapely.LineString([(1, 2, 3), (5, 6, 7)])  # LINESTRING Z (1 2 3, 5 6 7)
shapely.LineString([(1, 2, 3, 4), (5, 6, 7, 8)])  # LINESTRING ZM (1 2 3 4, 5 6 7 8)
shapely.LineString([(1, 2, None, 4), (5, 6, None, 8)])  # LINESTRING M (1 2 4, 5 6 8)
shapely.LineString([(1, 2, 4), (5, 6, 8)], some_option=x)  # LINESTRING M (1 2 4, 5 6 8)
```
the last example would specify a yet-to-be-decided keyword to signal that the third coordinates are M ordinates (the default would remain Z).

Opinions and example prototypes welcome!

**Comment by idanmiara** [comment]
> As @idanmiara noted, `include_z` is already used for `get_coordinates` in `src/coords.c`, so it would be consistent to add `include_m` for retrieving coordinates.

Wouldn't it be easier to modify `include_z` to accept some `enum | bool` so we don't need to add `include_m` to so many functions? This would allow the `include_z` to properly so only the function that actually parse it would need to be changed.

> Also the Python signatures, e.g. [`shapely.points`](https://shapely.readthedocs.io/en/2.0.1/reference/shapely.points.html). This could be one of several things:
> 
> * non-breaking:
>   `points(coords, y=None, z=None, indices=None, out=None, m=None, **kwargs)`
> * breaking, possibly yielding unexpected results:
>   `points(coords, y=None, z=None, m=None, indices=None, out=None, **kwargs)`
> * breaking, but safer for users to identify fix:
>   `points(coords, y=None, *, z=None, m=None, indices=None, out=None, **kwargs)`
> 
> If we are introducing breaking changes for shapely 2.1, should warnings be introduced to the 2.0.* series?
> 

I think that we should choose a non-breaking change and at a later version (after we issue a warning ahead) to make all the parameter rename/reorder all at once, there are many relevant cases, just to name one example - `ops.transform` vs `shapely.transform`

**Comment by mwtoews** [comment]
Looking at low-level C and Cython functions, there are several functions that have the arguments `unsigned int dims` and assume it is 2 for XY and 3 for XYZ. But obviously this will need to change for 2.1. These functions are:

- `PyGEOS_CoordSeq_FromBuffer(...)` (see [src/c_api.h](https://github.com/shapely/shapely/blob/ab56a0947e7928273aa2f4e18b8062635330ed16/src/c_api.h#L40-L42) and [shapely/_pygeos_api.pxd](https://github.com/shapely/shapely/blob/ab56a0947e7928273aa2f4e18b8062635330ed16/shapely/_pygeos_api.pxd#L55-L58))
- `coordseq_from_buffer(...)` (see [src/geos.h](https://github.com/shapely/shapely/blob/ab56a0947e7928273aa2f4e18b8062635330ed16/src/geos.h#L199-L204))
- `coordseq_to_buffer(...)` (see [src/geos.h](https://github.com/shapely/shapely/blob/ab56a0947e7928273aa2f4e18b8062635330ed16/src/geos.h#L205-L206))
- several `force_dims_*(...)` functions (see [src/geos.c](https://github.com/shapely/shapely/blob/ab56a0947e7928273aa2f4e18b8062635330ed16/src/geos.c))

Some options to replace the single `unsigned int dims` arg are:

- 2 args: `int hasZ, int hasM` (currently done for #1760)
- 2 args: `bool has_z, bool has_m`. Not only is this compact, but if needed `unsigned int = (unsigned int)has_z + (unsigned int)has_m + 2;` will be one of {2, 3, 4}.
- 1 arg as an enum like `CoordDim` as suggested above by @idanmiara. Unlike `GEOSGeomTypes`, GEOS does not have any built-in enums like "GEOSCoordTypes" to use.

**Comment by jorisvandenbossche** [comment]
For the low-level (private) coordseq-related APIs (`coordseq_from/to_buffer` et al), I would go with 2 args, given that this is consistent with the equivalent GEOS functions (and then I personally don't care whether we use the exact same as GEOS with `int hasZ` style, or `bool has_z`)

For the public Python-level methods like `transform`, that's another issue though. There it might indeed be useful to work with a single argumen. Although then we also need to use one with a different name than the current `include_z`, so that's not necessarily smoother compared to adding a `include_m`. But longer term a single argument might be the nicest option.

**Comment by tariqshihadah** [comment]
I'll add a major application for the `M` dimension--_linear referencing_. Linear referencing is a common data structure for linear asset data such as roadways, pipelines, etc. Much of this involves tabular operations, though some features include management of M-enabled linear geometries as well.

As a transportation data scientist this is quite common in my work. I've developed a [package](https://github.com/tariqshihadah/linref) called `linref` which provides some features related to this data structure. To this point I've used a bulky workaround, building on the `MultiLineString` geometry type and managing the `M` dimension in parallel.

Some of my collaborators on that project and I would be excited to have more complete support for `M` dimension data in `shapely` and would be happy to support the effort. The work-around module is a bit dated and limited but functional and may be a reference point for potential related features. In particular, the features of greatest value would be extending the existing linear geometry support for `interpolate` and `project` methods to account for the `M` dimension. Additionally, calibration of the `M` dimension.

Looks like there will be some opportunities to expand in this direction for `2.1` and beyond? Let me know if this is something I can support.

**Comment by mwtoews** [comment]
Hi all, I've been slow to get to these items for 2.1, but am now making some progress. I'm close to completing a PR for `get_coordinates()` that adds a `include_m=False` parameter, so perhaps I should describe the approach that I'm taking.

For low level-functions, I'm inserting `include_m`/`has_m` after `include_z`/`has_z`, even if there are other parameters like `return_index` that  follow. For public functions like [shapely.get_coordinates()](https://shapely.readthedocs.io/en/2.0.7/reference/shapely.get_coordinates.html) I'm appending the signature to have `, *, include_m`, which forces keyword-only for `include_m` and doesn't break positional `return_index`. But for shapely 2.1 I'm also raising a DeprecationWarning if `include_z` or `return_index` are passed as positional arguments (I have a nifty decorator to do this). Then for shapely 2.2 these can all be forced as keyword arguments and re-ordered as `include_z`, `include_m`, `return_index` (same as low-level function changes for shapely 2.1). This might make more sense when I post the PR, anticipated in a few days.

**Comment by jorisvandenbossche** [comment]
@mwtoews how much would you still like to include in 2.1? (`set_coordinates` / `transform`?) 
I think it is already great that we support accessing the coordinates, and would like to release 2.1 shortly ..

**Comment by mwtoews** [comment]
@jorisvandenbossche it's difficult to estimate how long these enhancements might take. There are also three geometry [creation functions](https://github.com/shapely/shapely/blob/250b9d8b22e1d58180964d3bb44bd9314c5e1d5a/shapely/creation.py) (`points` , `linestrings`, `linearrings`) that could add a `m=` parameter, so this adds up a further. Maybe see how it's going after a week?

A fallback option (that will take much less time to prepare) is to deprecate some positional arguments, like `indices` and `out` in preparation to make them keyword-only. This fallback option would imply these creation features to wait for shapely 2.2.

**Comment by jorisvandenbossche** [comment]
For those creation functions it will also be a question how to specify the dimensionality if you don't pass separate arrays for x/y/z/m, but you pass a single 2D array to `points()` for example. 

I know it is in this case easy for me to say as it is not my PR .. but maybe we should focus on 2.1 to say that we now support _accessing_ M values? And then support for M values in the various creation functions (except for WKT/WKB, which already works if you have recent GEOS) can be a new feature for 2.2. 
(and 2.2 does not have to take more than 2 years like 2.1 .. )

**Comment by jorisvandenbossche** [comment]
> I'm starting to agree with you that shapely 2.1 can be about accessing M coordinates, and that creating M values can wait a bit longer for shapely 2.2.

Then moving this issue to the 2.2 milestone for those further improvements
